### ISSUE TITLE
Question about the logprobs of the policy-generated sentences in PPO trainer

### ISSUE THREAD

**Issue by yanghh2000** [huggingface/trl#2358]
### System Info

trl==0.12.0

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

In the implementation of PPO trainer, the logits of the policy-generated sentences are inherited from outputs of the model's geneate(), i.e., output.scores. However, the returned logit scores are processed or filtered with the transformers's LogitsProcessor (e.g., Temperature, Top-k, Top-p), meaning they are not the original logits from model.forward(). 

```python
with unwrap_model_for_generation(model, self.accelerator) as unwrapped_model:
  query_responses, logitss = batch_generation(
      unwrapped_model.policy,
      queries,
      args.local_rollout_forward_batch_size,
      processing_class.pad_token_id,
      generation_config,
  )


for i in range(0, queries.shape[0], args.local_rollout_forward_batch_size):
  query = queries[i : i + args.local_rollout_forward_batch_size]
  query_response = query_responses[i : i + args.local_rollout_forward_batch_size]
  response = query_response[:, context_length:]
  logits = logitss[i : i + args.local_rollout_forward_batch_size]
  all_logprob = F.log_softmax(logits, dim=-1)
  logprob = torch.gather(all_logprob, 2, response.unsqueeze(-1)).squeeze(-1)
  del logits, all_logprob
  torch.cuda.empty_cache()
```

### Expected behavior

I think this can have an impact on ppo training as it may ignore some tokens's probility. And the possible solution is that we should use the policy/ref model to forward again to recompute the logits over the sentence.

### Checklist

- [X] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [X] I have included my system information
- [X] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [X] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [X] Any traceback provided is complete

**Comment by yaof20** [comment]
I am also not sure if the raw_logits is the same as the `logitss` returned by the `batch_generation`, as it is process by `LogitsProcessor`. (most likely they are different).

There is one possible solution -- in the `generate` function, you can specify `output_logits=True`, which I assume will return you the raw logits. 

Correct me if I am wrong.

**Comment by qgallouedec** [comment]
Hey @yanghh2000, sorry for the very slow reply here. Catching up on the PPO backlog.

Good question, but the current behavior is intentional and consistent. Two things:
1. PPO disables the top-k/top-p warpers in generation (`top_k=0.0`, `top_p=1.0`, see https://github.com/huggingface/trl/blob/40f670fc06007daa3d4824198ed47e477cb1a2cd/trl/experimental/ppo/ppo_trainer.py#L622-L625), so `output.scores` is never `-inf`-masked, the only active LogitsProcessor is the temperature warper, i.e. `scores = raw_logits / temperature`.
2. That temperature scaling is matched everywhere: the rollout logprobs use those tempered scores (https://github.com/huggingface/trl/blob/40f670fc06007daa3d4824198ed47e477cb1a2cd/trl/experimental/ppo/ppo_trainer.py#L703 and both the ref logprobs and the new-policy logprobs from the forward pass are manually divided by the same temperature before `log_softmax` https://github.com/huggingface/trl/blob/40f670fc06007daa3d4824198ed47e477cb1a2cd/trl/experimental/ppo/ppo_trainer.py#L713 and https://github.com/huggingface/trl/blob/40f670fc06007daa3d4824198ed47e477cb1a2cd/trl/experimental/ppo/ppo_trainer.py#L822

So all three logprob terms live on the same sampling distribution and the PPO ratio is consistent. Switching to `output_logits=True` (raw logits) would actually *break* this consistency, since the rollout term would be untempered while the forward-pass terms are tempered. 

Note PPO now lives under `trl.experimental.ppo`.
